packages/react/src/template.tsx

```diff
@@ -47,11 +47,31 @@ export type ActionDataState<T = any> = {
   $: 'action'
   name: T
 }
+export type ResponseDataState = {
+  $: 'response'
+  payload: JSONValue
+  statusCode: number
+  ok: boolean
+  actions: ActionDataState[]
+}
+export type HandlerDataState = {
+  $: 'handler'
+  state: Record<string, StateDataState>
+}
+export type HandlerReturnType = ResponseDataState | JSONValue | void
+export type HandlerFunction<T = any> = (args: T) => Promise<HandlerReturnType> | HandlerReturnType
 export type EventDataState = {
   $: 'event'
+  handler?: HandlerDataState
```

Comment:
maybe we should not have handler on the event. instead we could have:

```
type ClientState = EventState | ...
type HandlerState = { $: 'handler', handler: fn... }
type ServerState = ClientState | HandlerState
```

Then the system will automatically convert from ServerState to ClientState by converting HandlerState into EventState

Reply:
Could you elaborate a bit more on your example? I am not sure I got it right!

To further explain, the idea here was to scope changes to `fn`, without affecting the rest of the logic. So, before this PR, `event` only accepted a function `event(() => {})`. 

After this PR, `event` also takes `HandlerDataState` that can be instantiated with `withState`, e.g. `event(withState({}, () => {}))`, which allows us to modify the behavior of the function called. 

You can still use `event` tho to specify local actions, timeout etc., everything that you'd normally do anyway.

Reply:
I have removed this